README.md
xcinfo
======

*xcinfo* provides information of X cursor including, also the output:

    x y sw sh cw ch cx cy

where:

 * `x`, `y` : cursor position
 * `sw`, `sh`: screen resolution
 * `cw`, `ch`: cursor image size
 * `cx`, `cy`: cursor hotspot position

Compilation
-----------

    $ make

Installation
------------

    # By default, to install to `/usr/local`, run:
    $ make install

    # Or to `/usr`:
    $ make install PREFIX=/usr

    # Or to your home:
    $ make install PREFIX=$HOME

To uninstall, use `uninstall` target with `PREFIX` if supplied during installation.

Usage
-----

This code is written intentedly being used in shell script, e.g.

    :::bash
    read x y sw sh cw ch cx cy <<< "$(xcinfo)"

You can discard unneeded outputs, e.g.

    :::bash
    read x y sw sh _ <<< "$(xcinfo)"

### Options

`-i N`: querying every N microseconds.

Examples
--------

### bzen2.sh

A dzen2 which bounches, simply run:

    bzen2.sh message here, blah blah blah...

You can watch a screenshot in this [blog post][bzen2-post].

[bzen2-post]: http://blog.yjl.im/2011/09/bzen2-dzen2-which-bounces.html

### mouse trail

Scripts to record and render mouse trail as a SVG file, for example:

    $ mousetrack-record.sh mouse.pos
    Press Q to stop recording
    $ mousetrack-svg.sh mouse.pos mouse.svg

    # Converting to PNG format
    $ inkscape mouse.svg --export-background=FFF --export-png=mouse.png

    # or simply
    $ inkscape mouse.svg -b FFF -e mouse.png

### snap.sh

A script snaps a screenshot when mouse has movements every 5 minutes. It requires xsnap program to grab a screenshot.

### heatmap.sh

Drawing heatmap of X cursor onto terminal window size. Here is a [video](http://youtu.be/4CsI2qG1h1Q).

License
-------

This code is placed in Public Domain.


CHANGES.md
CHANGES
=======

## Development

## Version 0.3.0 (2013-10-19T10:34:43Z)

* add option parsing
* add -i for querying cursor by specified interval in microseconds
* add heatmap example

## Version 0.2 (2012-04-23T23:16:03Z)

  * Add snap and mouse trail examples

## Version 0.1 (2012-04-23T19:41:18Z)

The C code in this version is the same when it's hosted on GitHub.


xcinfo.c
/**
 * Written by Yu-Jie Lin on 9/11/2011
 * Public Domain
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <X11/Xlib.h>
#include <X11/extensions/Xfixes.h>


#define VERSION "0.3.0"


void
print_usage(FILE *out, char *cmd)
{
  fprintf(out, "Usage: %s [-v | -h] [-i usecs]\n", cmd);
}


void
print_help()
{
  printf("Options:\n\
\n\
-i      querying interval in microseconds\n\
-v      version information\n\
-h      help message\n\
\n\
Output is:\n\
\n\
  \"x y sw sh cw ch cx cy\"\n\
\n\
x , y  : cursor position\n\
sw, sh : screen resolution\n\
cw, ch : cursor image size\n\
cx, cy : cursor hotspot position\n\
\n\
Written intentedly being used in shell script, e.g.\n\
\n\
  read x y sw sh cw ch cx cy <<< \"$(xcinfo)\"\n\
\n\
You can discard unneeded outputs, e.g.\n\
\n\
  read x y sw sh _ <<< \"$(xcinfo)\"\n\
\n\
");
}


int
main(int argc, char **argv)
{
  int opt;
  useconds_t interval;
  int scr;
  int sw, sh;
  int px, py;
  Display *dpy;
  Window w, rt;
  XFixesCursorImage *ci;

  interval = 0;
  while ((opt = getopt(argc, argv, "i:vh")) != -1) {
    switch (opt) {
    case 'i':
      interval = atoi(optarg);
      break;
    case 'v':
      printf(VERSION "\n");
      exit(EXIT_SUCCESS);
    case 'h':
      print_usage(stdout, argv[0]);
      printf("\n");
      print_help();
      exit(EXIT_SUCCESS);
    default:
      print_usage(stderr, argv[0]);
      exit(EXIT_FAILURE);
    }
  }

  dpy = XOpenDisplay(NULL);
  scr = DefaultScreen(dpy);
  rt  = RootWindow   (dpy, scr);

  do {
    sw  = DisplayWidth (dpy, scr);
    sh  = DisplayHeight(dpy, scr);

    ci = XFixesGetCursorImage(dpy);
    printf("%d %d %d %d %d %d %d %d\n",
           ci->x, ci->y,
           sw, sh,
           ci->width, ci->height,
           ci->xhot,  ci->yhot);
    fflush(stdout);
    XFree(ci);
  } while (interval && !usleep(interval));

  XCloseDisplay(dpy);
}


examples/snap.sh
#!/bin/bash
# Snapping a picture when mouse has movements every 5 minutes
# Copyright (c) 2012 Yu-Jie Lin
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
# of the Software, and to permit persons to whom the Software is furnished to do
# so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

XCINFO=xcinfo

if ! type $XCINFO &>/dev/null; then
  if [[ -x ../xcinfo ]]; then
    XCINFO=../xcinfo
  else
    echo "Cannot find executable xcinfo." >&2
    exit 1
  fi
fi


# Get screen width and height
read _ _ SW SH _ <<< "$($XCINFO)"

while :; do
  read x y <<< "$($XCINFO)"
  POS="$x $y"
  if [[ $POS != $OLD_POS ]]; then
    filename="/tmp/snap-$(date +'%Y-%m-%d--%H:%M:%S.png')"
    xsnap -file "$filename" -region "${SW}x${SH}+0+0"
    OLD_POS="$POS"
    echo "$filename SNAP!"
  fi
  sleep 5m
done


examples/mousetrack-svg.sh
#!/bin/bash

XCINFO=xcinfo
IN_FILE=/dev/stdin
OUT_FILE=/dev/stdout

if ! type $XCINFO &>/dev/null; then
  if [[ -x ../xcinfo ]]; then
    XCINFO=../xcinfo
  else
    echo "Cannot find executable xcinfo." >&2
    exit 1
  fi
fi

if (( $# != 2 )); then
  echo "Usage: $0 <in_file> <out_svg>" >&2
  exit 1
fi

if [[ "$1" != - ]]; then
  IN_FILE="$1"
fi

if [[ "$2" != - ]]; then
  OUT_FILE="$2"
fi

# Get screen width and height
read _ _ SW SH _ <<< "$($XCINFO)"

echo '<?xml version="1.0"?>  
<svg width="1680" height="1050"
     xmlns="http://www.w3.org/2000/svg" version="1.1"  
     xmlns:xlink="http://www.w3.org/1999/xlink">  
<defs>
  <radialGradient id="g">  
    <stop offset="0" stop-color="#000" stop-opacity="1"/>
    <stop offset="1" stop-color="#000" stop-opacity="0"/>
  </radialGradient>  
</defs>' > "$OUT_FILE"

# Use uniq to remove repeating records because the inactivity of mouse
while read x y; do
  echo '<circle fill="url(#g)" r="10" cx="'$x'" cy="'$y'"/>'
done < <(uniq "$IN_FILE") >> "$OUT_FILE"

echo '</svg>' >> "$OUT_FILE"


examples/heatmap.sh
#!/bin/bash
# X cursor heatmap in terminal
# Copyright (c) 2013 Yu-Jie Lin
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
# of the Software, and to permit persons to whom the Software is furnished to do
# so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

# Violet -> Red
RB=(129 92 55 56 21 26 31 36 41 46 82 118 154 190 226 220 214 208 202 196)
RBN=${#RB[@]}

XCINFO=xcinfo

if ! type $XCINFO &>/dev/null; then
  if [[ -x ../xcinfo ]]; then
    XCINFO=../xcinfo
  else
    echo "Cannot find executable xcinfo." >&2
    exit 1
  fi
fi


# Get screen width and height
read _ _ SW SH _ <<< "$($XCINFO)"
# Get terminal size
TW=$(tput cols)
TH=$(tput lines)

draw() {
  local H=$1 X=$2 Y=$3
  if ((H)); then
    ((RBIDX = H * (RBN - 1) / 100))
    C=${RB[RBIDX]}
  else
    C=16
  fi
  echo -ne "\e[$((Y + 1));$((X + 1))H\e[48;5;${C}m "
}


HM=()
for ((x=0; x<TW; x++)); do
  for ((y=0; y<TH; y++)); do
    ((IDX = x + y * TW, HM[IDX] = 0))
    draw ${HM[IDX]} $x $y
  done
done

RC=0
"$XCINFO" -i 10000 |
while read XX XY _; do
  ((TX = XX * TW / SW, TY = XY * TH / SH))
  ((IDX = TX + TY * TW, HM[IDX] = HM[IDX] + 5))
  ((HM[IDX] > 100)) && HM[IDX]=100
  draw ${HM[IDX]} $TX $TY
  if ((++RC >= 100)); then
    for ((x=0; x<TW; x++)); do
      for ((y=0; y<TH; y++)); do
        ((IDX = x + y * TW))
        if ((HM[IDX])); then
          ((HM[IDX]--))
          draw ${HM[IDX]} $x $y
        fi
      done
    done
    RC=0
  fi
done
